RedBlackTree: Error at left

```golang
DEBUG[09-09|15:55:28] After matching, order (unmatched part) is now back to asks tree order="&{Quantity:+20000000000000000 Price:+20047000000000000000000 ExchangeAddress:[13 58 177 75 186 211 217 159 66 3 189 122 17 172 185 72 130 5 14 126] UserAddress:[23 242 190 215 16 186 80 237 39 174 165 47 196 189 123 218 94 212 160 55] BaseToken:[77 126 162 206 148 146 22 214 177 32 243 170 16 22 65 115 97 90 43 108] QuoteToken:[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1] Status:New Side:SELL Type:LO Hash:[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 49 48 50] Signature:0xc00fd29680 FilledAmount:+0 Nonce:+102 MakeFee:+4000000000000000 TakeFee:+4000000000000000 PairName:BTC/TOMO CreatedAt:1568019327 UpdatedAt:1568019327 OrderID:93 NextOrder:[] PrevOrder:[] OrderList:[] Key:}"
DEBUG[09-09|15:55:28] create price list                        detail=20047000000000000000000
DEBUG[09-09|15:55:28] Save orderlist                           key=6274632f746f6d6fa8e3c169521e72e3bee3c02630a1667a3ce8dfa0598bc996381674b6fe5811a8 value="[248 79 160 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 160 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 128 128 138 4 62 192 3 30 113 31 220 0 0]"
DEBUG[09-09|15:55:28] Key not found                            key=6274632f746f6d6fa8e3c169521e72e3bee3c02630a1667a3ce8dfa0598bc99653d7e21e4d2011a8 err="leveldb: not found"
ERROR[09-09|15:55:28] Error at left                            err="leveldb: not found"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0xb60055]

goroutine 32 [running]:
github.com/ethereum/go-ethereum/tomox.(*Tree).Put(0xc01c0b8740, 0xc01c0e21b0, 0x28, 0x30, 0xc01c0dc300, 0x51, 0x51, 0x1, 0x421b78, 0x30)
        /home/nguyen/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/redblacktree.go:79 +0x575
github.com/ethereum/go-ethereum/tomox.(*OrderTree).SaveOrderList(0xc00feeae60, 0xc01c0d2de0, 0x1, 0x1b33278, 0xc01c0d6c40)
        /home/nguyen/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/ordertree.go:183 +0x475
github.com/ethereum/go-ethereum/tomox.(*OrderList).Save(...)
        /home/nguyen/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/orderlist.go:161
github.com/ethereum/go-ethereum/tomox.(*OrderTree).CreatePrice(0xc00feeae60, 0xc00fdc6680, 0x1, 0x4)
        /home/nguyen/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/ordertree.go:167 +0x5b
github.com/ethereum/go-ethereum/tomox.(*OrderTree).InsertOrder(0xc00feeae60, 0xc00fc90b00, 0x1, 0x4, 0xc004d80c20)
        /home/nguyen/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/ordertree.go:237 +0x154
github.com/ethereum/go-ethereum/tomox.(*OrderBook).processLimitOrder(0xc00feeadc0, 0xc00fc90b00, 0x1, 0x1, 0xc004d80f10, 0x2, 0x2, 0x1, 0xc00ff64f01, 0xc01c0da7f8)
        /home/nguyen/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/orderbook.go:320 +0x828
github.com/ethereum/go-ethereum/tomox.(*OrderBook).ProcessOrder(0xc00feeadc0, 0xc00fc90b00, 0x3c7a66a13026c001, 0x1, 0xca81168af107982, 0x0, 0x0, 0x2, 0x0, 0x0)
        /home/nguyen/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/orderbook.go:350 +0x163
github.com/ethereum/go-ethereum/tomox.(*TomoX).ProcessOrderPending(0xc0000e2120, 0x124ae26, 0x1e, 0x4)
        /home/nguyen/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/tomox/tomox.go:866 +0x84c
github.com/ethereum/go-ethereum/miner.(*worker).commitNewWork(0xc0002f4c80)
        /home/nguyen/go/src/github.com/ethereum/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/miner/worker.go:617 +0x1c28
```

--- a/tomox/filter.go
+++ b/tomox/filter.go
@@ -155,7 +155,7 @@ func (fs *Filters) NotifyWatchers(env *Envelope, p2pMessage bool) {
 		}
 
 		if match && msg != nil {
-			log.Debug("processing message: decrypted", "hash", env.Hash().Hex())
+			log.Trace("processing message: decrypted", "hash", env.Hash().Hex())
 			watcher.Trigger(msg)
 		}
 	}

--- a/tomox/node.go
+++ b/tomox/node.go
@@ -133,7 +133,7 @@ func (node *Node) Right(tree *Tree, dryrun bool, blockHash common.Hash) *Node {
 	key := node.RightKey()
 	newNode, err := tree.GetNode(key,dryrun, blockHash)
 	if err != nil {
-		fmt.Println(err)
+		log.Error("Error at right", "err", err)
 	}
 	return newNode
 }

--- a/tomox/peer.go
+++ b/tomox/peer.go
@@ -102,7 +102,7 @@ func (peer *Peer) update() {
 
 		case <-transmit.C:
 			if err := peer.broadcast(); err != nil {
-				log.Debug("broadcast failed", "reason", err, "peer", peer.ID())
+				log.Trace("broadcast failed", "reason", err, "peer", peer.ID())
 				return
 			}
 
@@ -161,7 +161,7 @@ func (peer *Peer) broadcast() error {
 			peer.mark(e)
 		}
 
-		log.Debug("broadcast", "num. messages", len(bundle))
+		log.Trace("broadcast", "num. messages", len(bundle))
 	}
 	return nil
 }

--- a/tomox/redblacktree.go
+++ b/tomox/redblacktree.go
@@ -79,7 +79,6 @@ func (tree *Tree) Put(key []byte, value []byte, dryrun bool, blockHash common.Ha
 			compare := tree.Comparator(key, node.Key)
 			switch {
 			case compare == 0:
-
 				node.Item.Value = value
 				tree.Save(node, dryrun, blockHash)
 				return nil
@@ -95,7 +94,6 @@ func (tree *Tree) Put(key []byte, value []byte, dryrun bool, blockHash common.Ha
 					node = node.Left(tree, dryrun, blockHash)
 				}
 			case compare > 0:
-
 				if tree.IsEmptyKey(node.RightKey()) {
 					node.RightKey(key)
 					tree.Save(node, dryrun, blockHash)
@@ -200,6 +198,9 @@ func (tree *Tree) Remove(key []byte, dryrun bool, blockHash common.Hash) {
 					log.Error("Parent doesn't have node as child", "node.Key", hex.EncodeToString(node.Key))
 					return
 				}
+			} else {
+				// this node is root
+				tree.rootKey = EmptyKey()
 			}
 			tree.deleteNode(node, dryrun, blockHash)
 			log.Debug("Removed node with child = nil", "node", hex.EncodeToString(node.Key), "all keys", tree.KeysinString(true, blockHash))

--- a/tomox/tomox_test.go
+++ b/tomox/tomox_test.go
@@ -18,11 +18,11 @@ func buildOrder(nonce *big.Int) *OrderItem {
 	v := []byte(string(rand.Intn(999)))
 	lstBuySell := []string{"BUY", "SELL"}
 	order := &OrderItem{
-		Quantity:        new(big.Int).SetUint64(uint64(rand.Intn(10)) * 1000000000000000000),
-		Price:           new(big.Int).SetUint64(uint64(rand.Intn(10)) * 100000000000000000),
+		Quantity:        new(big.Int).SetUint64(uint64(rand.Intn(9)+1) * 1000000000000000000),
+		Price:           new(big.Int).SetUint64(uint64(rand.Intn(9)+1) * 100000000000000000),
 		//Quantity: new(big.Int).SetUint64(uint64(5) * 1000000000000000000),
 		//Price:           new(big.Int).SetUint64(uint64(2) * 100000000000000000),
-		ExchangeAddress: common.HexToAddress("0x0342d186212b04E69eA682b3bed8e232b6b3361a"),
+		ExchangeAddress: common.HexToAddress("0x0D3ab14BBaD3D99F4203bd7a11aCB94882050E7e"),
 		UserAddress:     common.HexToAddress("0x9ca1514E3Dc4059C29a1608AE3a3E3fd35900888"),
 		BaseToken:       common.HexToAddress("0x4d7eA2cE949216D6b120f3AA10164173615A2b6C"),
 		QuoteToken:      common.HexToAddress("0xC2fa1BA90b15E3612E0067A0020192938784D9C5"),
@@ -30,7 +30,7 @@ func buildOrder(nonce *big.Int) *OrderItem {
 		Side:            lstBuySell[rand.Int()%len(lstBuySell)],
 		//Side: "SELL",
 		Type:            Limit,
-		PairName:        "0x4d7eA2cE949216D6b120f3AA10164173615A2b6C" + "::" + "0xC2fa1BA90b15E3612E0067A0020192938784D9C5",
+		PairName:        "BTC/ETH",
 		//Hash:            common.StringToHash("0xdc842ea4a239d1a4e56f1e7ba31aab5a307cb643a9f5b89f972f2f5f0d1e7587"),
 		Hash: common.StringToHash(nonce.String()),
 		Signature: &Signature{
@@ -86,9 +86,9 @@ func TestCreate10Orders(t *testing.T) {
 	// disable this test in travis CI
 	t.SkipNow()
 
-	for i := 1; i <= 100; i++ {
+	for i := 1001; i <= 2000; i++ {
 		testCreateOrder(t, new(big.Int).SetUint64(uint64(i)))
-		time.Sleep(1 * time.Second)
+		time.Sleep(100 * time.Millisecond)
 	}
 }
 

